File: backend/src/main/java/dev/voroby/client/dto/NewMessage.java
package dev.voroby.client.dto;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = NewMessage.TextMessage.class, name = NewMessage.TEXT_MESSAGE)
})
public sealed interface NewMessage {

    String TEXT_MESSAGE = "TextMessage";

    record TextMessage(
            long chatId,
            String text
    ) implements NewMessage {
    }

}

File: desktop/src/main/kotlin/chat/dto/NewMessage.kt
package chat.dto

sealed interface NewMessage {

    val type: MessageType

    val chatId: Long

    data class TextMessage(
        override val type: MessageType = MessageType.TextMessage,
        override val chatId: Long,
        val text: String
    ) : NewMessage
}

enum class MessageType { TextMessage }
